### 정확도가 매우 중요한 경우 decimal을 사용하라



<br>


- 정밀한 계산이 필요할 때 Decimal 클래스(십진수로 그대로 표현)

In [30]:
rate = 1.45
seconds = 3*60 + 42
cost = rate * seconds / 60
cost

5.364999999999999

- 부동소수점을 통해 값을 표현하기 때문에 5.365는 정확하게 나타낼 수 없음
- 0.365를 2의 거듭제곱꼴로 완전히 표현할 수 없으므로

#### decimal 객체로 해결가능

In [31]:
from decimal import *

In [32]:
rate = Decimal('1.45')
seconds = Decimal(3*60 + 42)
cost = rate * seconds / Decimal(60)
cost

Decimal('5.365')

In [33]:
Decimal('1.45')

Decimal('1.45')

In [34]:
Decimal(1.45)

Decimal('1.4499999999999999555910790149937383830547332763671875')

Decimal 안의 값이 float객체로 들어가면 2진수 자체로 표현된 수가 나옴

but 정수는 상관 없음

In [35]:
Decimal(456)

Decimal('456')

In [36]:
rate = Decimal('0.05')
seconds = Decimal(5)
small_cost = rate * seconds / Decimal(60)
small_cost 

Decimal('0.004166666666666666666666666667')

In [38]:
rounded = cost.quantize(Decimal('0.01'), rounding=ROUND_UP) #소수 둘째자리에서 올림
rounded

Decimal('5.37')

In [40]:
rounded = small_cost.quantize(Decimal('0.01'), rounding=ROUND_HALF_UP) #소수 둘째자리에서 반올림
rounded

Decimal('0.00')

### 한계


<br>

- 10진수로도 순환소수가 나오는 경우 => 정밀하지 않음




In [44]:
a = Decimal('1') / Decimal('3') * 3
a

Decimal('0.9999999999999999999999999999')

### Fraction 클래스 사용

In [45]:
from fractions import *

In [49]:
a = Fraction(1, 3) * 3
a

Fraction(1, 1)